### PR TITLE
WIP: Code Actions

### PR DESCRIPTION
**Pull request by ctolkien**
Not finished, PR more for visibility at this point

### CODE REVIEW

File: keymaps/omnisharp-atom.cson [hunk 1]
```diff
@@ -31,6 +31,7 @@
   'f1': 'omnisharp-atom:type-lookup'
   'ctrl-alt-up': 'omnisharp-atom:navigate-up'
   'ctrl-alt-down': 'omnisharp-atom:navigate-down'
+  'ctrl-.': 'omnisharp-atom:get-code-actions'
```
Comment (nosami) [inline keymaps/omnisharp-atom.cson]: needs alt+enter too ;)

File: lib/omnisharp-atom/features/code-action.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+class CodeCheck {
+
+    view;
+
+    constructor(private atomSharper: typeof OmniSharpAtom) {
+        this.atomSharper = atomSharper;
+    }
+    public activate() {
+        atom.commands.add('atom-workspace', "omnisharp-atom:get-code-actions", () => {
```
Comment (nosami) [inline lib/omnisharp-atom/features/code-action.ts]: we just added OmniSharpAtom.addCommand, so you can do `OmniSharpAtom.addCommand("omnisharp-atom:get-code-actions", () => {`

This takes care of making sure that the command will only run for a c# file.

**Comment by nosami** [comment]
I've been looking into how to use the command palette some more because I want to use it for some features.

Atom-typescript uses a generic class for this https://github.com/TypeStrong/atom-typescript/blob/d5ca888a6c498f58e57ac8170836a391c029cef8/lib/main/atom/views/simpleSelectionView.ts

Here's an example of how it's used https://github.com/TypeStrong/atom-typescript/blob/43a7df7d4273fb01d4c4deae2d8facbeee027c43/lib/main/atom/commands/commands.ts#L282

I think we should do the same.

File: lib/omnisharp-atom/features/code-action.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            client.observeGetcodeactions.subscribe((data) => {
+
+                //pop ui to user.
+                this.view = new CodeActionsView(data.response.CodeActions, () => {
+                    //callback when an item is selected
+                    ClientManager.getClientForActiveEditor()
+                        .subscribe(client => {
+                            client.runcodeactionPromise(client.makeDataRequest<OmniSharp.Models.CodeActionRequest>(
+                                {
+                                    CodeAction : 0, //this is the "ID"(index?) of the code action to run...
```
Comment (ctolkien) [inline lib/omnisharp-atom/features/code-action.ts]: @nosami  - `getcodeactions` give us an array of strings for the available code actions, but `runcodeactions` wants a numeric identifier. Is this just the index of what what is returned from `getcodeactions`? (0 being first, then 1, etc.?)

Comment (nosami) [inline lib/omnisharp-atom/features/code-action.ts]: Yes. That's how it is right now.

We're planning on changing it at some point to use the ID of the fix instead.

File: lib/omnisharp-atom/features/code-action.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            client.observeRuncodeaction.subscribe((data) => {
+                this.applyAllChanges(data.response.Changes);
+            })
+        })
+
+    }
+
+    public applyAllChanges(changes: OmniSharp.Models.LinePositionSpanTextChange[]) {
+        //bug: if they have swapped editors since firing this request,
+        //it might apply the changes to the wrong editor?
```
Comment (nosami) [inline lib/omnisharp-atom/features/code-action.ts]: yes :)

**Comment by nosami** [comment]
Awesome! We can look at getting lightbulb or something else later.

**Comment by Mpdreamz** [comment]
w00t `SpacePen.SelectListView` works for now :)

**Pull request merged**